### PR TITLE
undo: add support for masks.

### PR DESCRIPTION
**Pull request by TurboGit**
Masks are now supported by the undo/redo circuitry as done for the
development data.

### CODE REVIEW

**Comment by TurboGit** [comment]
This is working, but it is painful as the changes are committed very often in darktable. Good as we do not loose edits, but difficult for the undo on mask to be useful. For example, create a mask on exposure, change the size with the mouse wheel, change the size of the feather with the mouse wheel... Do undo... You'll need lot of them to get rid of the mask.

We have the same issue when dragging curves in iop like "tone curve".

A solution has to be discussed.

**Comment by TurboGit** [comment]
I'd like to move forward on this (and #1419) which are really meant to work together. Can someone review this? I'd like to have this merged sooner than later to allow fixing potential issues before next major release.

File: src/develop/masks/masks.c [hunk 1]
```diff
@@ -35,6 +36,144 @@
...
+  dt_masks_form_t *form;
+} _masks_undo_data_t;
+
+static void _masks_write_form_db(dt_masks_form_t *form, dt_develop_t *dev);
+// write a form into the database
+
+static void _masks_write_forms_db(dt_develop_t *dev, gboolean undo);
+// write all masks form into the database. record an undo is undo is true
+
+static void _masks_free_form_list(GList *forms)
```
Comment (houz) [inline src/develop/masks/masks.c]: Why this function? Isn't that the same as `g_list_free_full(forms, dt_masks_free_form);`?

File: src/develop/masks/masks.c [hunk 2]
```diff
@@ -35,6 +36,144 @@
...
+{
+  GList *forms;
+  dt_masks_form_t *form;
+} _masks_undo_data_t;
+
+static void _masks_write_form_db(dt_masks_form_t *form, dt_develop_t *dev);
+// write a form into the database
+
+static void _masks_write_forms_db(dt_develop_t *dev, gboolean undo);
+// write all masks form into the database. record an undo is undo is true
```
Comment (houz) [inline src/develop/masks/masks.c]: `s/is/if/`?

Comment (TurboGit) [inline src/develop/masks/masks.c]: Right.

File: src/develop/masks/masks.c (hunk 1 above)
Comment (TurboGit) [inline src/develop/masks/masks.c]: An oversight. Now fixed (and rebased on master).

**Comment by TurboGit** [comment]
If there is no more objection I'll merge at the end of the week.

**Pull request merged**